### ISSUE TITLE
[FR] Multiple IPNs

### ISSUE THREAD

**Issue by Weissnix4711** [inventree/InvenTree#3374]
**Is your feature request the result of a bug?**
No

**Problem**
I need to use multiple different part numbering systems, which can overlap for some parts. Ie, each IPN is unique, but a part may have multiple IPNs.

**Suggested solution**
Would it be possible to add the ability to create multiple internal part numbers, per part. Ideally, these would also have individual regex patterns in the settings, though that's not entirely necessary.

**Examples of other systems**
Not sure tbh

**Do you want to develop this?**
I would like to, but I don't really have the time right now, hence I'm posting this FR. But I'm hoping it should be too difficult to implement.

**Comment by matmair** [comment]
Hi there @Weissnix4711 ! Sounds a bit like the problem described here: #3369 .
The time problem sounds familiar, not sure if and how fast the core team has time to do anything on this.

**Comment by Weissnix4711** [comment]
Not quite the same issue I'm afraid. I have various internal numbering systems I have to deal with somehow, which have nothing to do with the supplier or manufacturer.

Anyway, don't worry about it too much. I get that you guys have many more important things to be dealing with. I'll try my best to throw something together myself.

**Comment by matmair** [comment]
@Weisnix4711 I think long-term this is an interesting topic but for now there are ~40 other things that I will spend my focus on.
Sweat username btw. - are you a fan of german cologne by any chance?

**Comment by SchrodingersGat** [comment]
@Weissnix4711 can you use comma-separated values in the existing IPN field to achieve what you want?

**Comment by Weissnix4711** [comment]
> @Weissnix4711 can you use comma-separated values in the existing IPN field to achieve what you want?

Not really, I'm afraid. Inventree still handles that as one long IPN, which means trying to get info about a part via the api is annoying when only one of the IPNs is known. Sure, I could use regex to search for a part's IPN which includes a certain substring, but it's more of a hassle than before. Also, I can't access the part via `/part/IPN1234567/` unless I know all of its IPN values.

Eg, a part has two IPNs, `IC-00013725-00A` and `943WFWFQ40D8PC5K`. If I were to enter that into Inventree as `IC-00013725-00A,943WFWFQ40D8PC5K` then I cannot access the part at `/part/IC-00013725-00A/` or `/part/943WFWFQ40D8PC5K/`. But only using the full string.

There's probably also a dozen other places where this comma separated approach would fall apart, but can't think rn.

That can be worked around, but I would much rather have native support for multiple IPNs. At least in the long term.

**Comment by Weissnix4711** [comment]
> I think long-term this is an interesting topic but for now there are ~40 other things that I will spend my focus on.

Ofc, ofc.

> Sweat username btw

Thanks. A long time ago, I was trying to get the username 'Weissnix' on a platform. Can't even remember what exactly it was for. That was taken, so decided to throw a random number on the end. First thing that came to mind. Anyway, that's now stuck with me and all my accounts pretty much the same.

**Comment by SchrodingersGat** [comment]
@Weissnix4711 sounds like an "edge case" for what is currently supported. I don't see a particularly clean way of implementing this in the short term.

One thing you might consider is overloading the "supplier part" functionality, creating multiple "supplier parts" which really just are internal parts? i.e. the "supplier" in this case is just yourself?

**Comment by matmair** [comment]
> First thing that came to mind. Anyway, that's now stuck with me and all my accounts pretty much the same.

The number combination has several meanings in some regions of Germany and Austria 😇.

**Comment by miggland** [comment]
> Eg, a part has two IPNs, `IC-00013725-00A` and `943WFWFQ40D8PC5K`. If I were to enter that into Inventree as `IC-00013725-00A,943WFWFQ40D8PC5K` then I cannot access the part at `/part/IC-00013725-00A/` or `/part/943WFWFQ40D8PC5K/`. But only using the full string.

Wait, how does that work? I've thought about being able to do that a few times, but for me it only works using the part PK, and not the IPN?!

**Comment by Weissnix4711** [comment]
> Wait, how does that work? I've thought about being able to do that a few times, but for me it only works using the part PK, and not the IPN?!

It's definitely a thing. For example, https://demo.inventree.org/part/TB3/

**Comment by miggland** [comment]
OK, I can see that that works.. but not on my system! I get a page not found. And I don't use template parts, so that's not it. Anything I have to activate to get it to work?

**Comment by Weissnix4711** [comment]
> Anything I have to activate to get it to work?

Not as far as I know. It works on  my server, which is pretty much 'vanilla'.

Have you added the trailing slash? Without, it will simply redirect to the Parts page. Does the IPN contain any special characters which might not work as a URL?

**Comment by miggland** [comment]
The trailing slash has caught me out before, so I did make sure I included it. If I don't include the slash, I see the part list page - if I include it, I get "page not found".

My IPNs are (nearly) exclusively 5-figure digits, sometimes with leading 0. Eg `/part/50200/` or `/part/90317/`.

Probably should open an issue about this, seems like something is not working as intended. Now I at least know that the possibility should exist!

**Comment by matmair** [comment]
This IPN-containing URL is not part of the core functionset and falls apart as soon as your part is not fitting the old IPN schema: `^\d*[a-zA-Z]{1,}\d*`. As long as there is at least one alphabetical character in the IPN it should work.

I would not bet on it working and use the pk/id. Only those are guaranteed to stay the same throughout the object life-cycle btw - Users can change IPNs.

**Comment by SchrodingersGat** [comment]
@Weissnix4711 still any interest in this? Otherwise we will close

**Comment by Weissnix4711** [comment]
Defo. I have to admit, I've been a bit lazy and haven't even bothered trying to implement it.

I've now just ditched some of the numbering systems I figured I probably wouldn't need. Would still be nice though in the long term to be able to add multiple IPNs per part, just in case.

**Comment by matmair** [comment]
Still relevant; the bot will trigger till we assign someone/a milestone

**Comment by wolflu05** [comment]
This one is still relevant.
